Explain how to work with subtree (#70654)
rust-lang/rust: CONTRIBUTING.md

Speaking of tests, Rust has a comprehensive test suite. More information about
it can be found [here][rctd].

### External Dependencies (subtree)

As a developer to this repository, you don't have to treat the following external projects
differently from other crates that are directly in this repo:

* none so far, see https://github.com/rust-lang/rust/issues/70651 for more info

They are just regular files and directories. This is in contrast to `submodule` dependencies
(see below for those).

#### Synchronizing a subtree

There are two synchronization directions: `subtree push` and `subtree pull`.

A `git subtree push -P src/tools/clippy`
takes all the changes that
happened to the copy in this repo and creates commits on the remote repo that match the local
changes (so every local commit that touched the subtree causes a commit on the remote repo).

A `git subtree pull -P src/tools/clippy` takes all changes since the last `subtree pull` from the clippy

> repo and creates a single commit in the rustc repo with all the changes.

Reviewer: Two things here:

1. From the [git documentation](), the syntax for these commands is
    ```
    'git subtree' pull  -P <prefix> <repository> <ref>
    'git subtree' push  -P <prefix> <repository> <ref>
    ```
    I think it is also important to note for `push`, that this shouldn't be done directly to the Clippy master branch, but to a new branch + PR, so it will get checked by the Clippy CI (formatting, integration tests, ...)
2. "creates a single commit in the rustc repo with all the changes." is not true for `git subtree pull`, is it?

Speaking of tests, Rust has a comprehensive test suite. More information about
it can be found [here][rctd].

### External Dependencies (subtree)

As a developer to this repository, you don't have to treat the following external projects
differently from other crates that are directly in this repo:

* none so far, see https://github.com/rust-lang/rust/issues/70651 for more info

They are just regular files and directories. This is in contrast to `submodule` dependencies
(see below for those).

If you want to synchronize or otherwise work with subtrees, install the `git subtree` command via

> instructions found at https://github.com/ingydotnet/git-subtree

Reviewer 2: Hmm, `git subtree` is already installed on my local copy of git (Ubuntu 18.04). Is it needed to install it from GitHub?

Reviewer 3: Also https://github.com/ingydotnet/git-subtree is was last updated 2011. It says "experimental" in the description. And it is a fork of https://github.com/matthoffman/git-subtree which is a fork of https://github.com/bibendi/git-subtree which is a fork of https://github.com/apenwarr/git-subtree. Is that really the one we want?^^

Reviewer 4: This is just a consequence of `s/subrepo/subtree/`, nothing to see here.
(The whole paragraph needs to be removed.)

Reviewer 3: Ah, `subtree` looks to be upstream these days: https://github.com/git/git/blob/master/contrib/subtree/git-subtree.txt

Speaking of tests, Rust has a comprehensive test suite. More information about
it can be found [here][rctd].

### External Dependencies (subtree)

As a developer to this repository, you don't have to treat the following external projects
differently from other crates that are directly in this repo:

* none so far, see https://github.com/rust-lang/rust/issues/70651 for more info

They are just regular files and directories. This is in contrast to `submodule` dependencies
(see below for those). Only tool authors will actually use any operations here.

#### Synchronizing a subtree

There are two synchronization directions: `subtree push` and `subtree pull`.

> `git subtree push -P src/tools/clippy https://github.com/rust-lang/rust-clippy.git`

Reviewer 3: Make make these triple-`` ` `` blocks?

takes all the changes that

> happened to the copy in this repo and creates commits on the remote repo that match the local

Reviewer 3: So push (and also pull) both create one commit "on the other side" for each commit "on this side"? There is no squashing in either direction?

Author: Yes, no squashing at all. In the pull direction we apparently even preserve commit hashes.

Reviewer 5: AFAICT, each commit is copied to the other side verbatim (even the hash is the same, at least for pull, not sure about push), then a single "subtree merge commit" is added which places the changes at the right prefix (when pulling) or removes the prefix (when pushing).

Reviewer 3: Sounds good. That might be worth mentioning in the docs as well.

Author: Not sure what is missing from the docs. I've stated that push commits get modified and pull commits get copied. Do you want me to rehash this in detail? Seems to me like this would be useless noise to the person looking for usage instructions. I'd just be moving `git subtree` documentation into our documentation in a trimmed and incomplete way.

Reviewer 3: Linking to well-written existing `git subtree` documentation is certainly also an option. :)

I just feel like one should be able to learn the results of the experiments that @eddyb described above by reading this document and following links.

Author: I understand that we need this information to make a decision, but I don't see how it helps us to have this documentation in the contribution guide.

Reviewer 3: I think it's good to understand the tool one is using, and this is a key part to understanding how `git subtree` works. It is the kind of knowledge you need to debug things when things get weird (and knowing git, they *will* get weird sometimes).

Author: I added more docs

Speaking of tests, Rust has a comprehensive test suite. More information about
it can be found [here][rctd].

### External Dependencies (subtree)

As a developer to this repository, you don't have to treat the following external projects
differently from other crates that are directly in this repo:

* none so far, see https://github.com/rust-lang/rust/issues/70651 for more info

They are just regular files and directories. This is in contrast to `submodule` dependencies
(see below for those). Only tool authors will actually use any operations here.

#### Synchronizing a subtree

There are two synchronization directions: `subtree push` and `subtree pull`.

```
git subtree push -P src/tools/clippy https://github.com/rust-lang/rust-clippy.git
```

takes all the changes that
happened to the copy in this repo and creates commits on the remote repo that match the local
changes (so every local commit that touched the subtree causes a commit on the remote repo).

```

> git subtree pull -P src/tools/clippy https://github.com/rust-lang/rust-clippy.git

Reviewer: With push and pull, also a `<ref>` has to be specified. There should be a note, that `push` should **never** aim at `master`, but at a new branch (either in a fork or the main repo) and to open a PR, so every sync is checked by the Clippy CI.

```suggestion
git subtree pull -P src/tools/clippy https://github.com/rust-lang/rust-clippy.git master
```

Reviewer 5: Also, `.git` suffixes are never required, and it's cleaner to omit them, especially in docs.

Speaking of tests, Rust has a comprehensive test suite. More information about
it can be found [here][rctd].

### External Dependencies (subtree)

As a developer to this repository, you don't have to treat the following external projects
differently from other crates that are directly in this repo:

* none so far, see https://github.com/rust-lang/rust/issues/70651 for more info

They are just regular files and directories. This is in contrast to `submodule` dependencies
(see below for those). Only tool authors will actually use any operations here.

#### Synchronizing a subtree

There are two synchronization directions: `subtree push` and `subtree pull`.

```
git subtree push -P src/tools/clippy git@github.com:your-github-name/rust-clippy rustup
```

takes all the changes that
happened to the copy in this repo and creates commits on the remote repo that match the local
changes. Every local commit that touched the subtree causes a commit on the remote repo, but is
modified to move the files from the specified directory to the tool repo root.

> Make sure to not pick the `master` branch, so you can open a normal PR to the tool to merge that

Reviewer 5: Also, we should just use some name like `sync-from-rust` in examples (elsewhere in this doc) and people can tweak that if they want different branch names.
I'd say `rustup` but that can be confusing in the new setup.

You can make PRs using the `master` branch of your fork anyway, so this paragraph may be unnecessary anyway.

subrepo push.

```
git subtree pull -P src/tools/clippy https://github.com/rust-lang/rust-clippy master
```

takes all changes since the last `subtree pull` from the tool repo
repo and adds these commits to the rustc repo + a merge commit that moves the tool changes into
the specified directory in the rust repository.

It is recommended that you always do a push first and get that merged to the tool master branch.
Then, when you do a pull, the merge works without conflicts.

> While definitely possible to resolve conflicts during a pull, you may have to redo the conflict

Reviewer 6: ```suggestion
While it's definitely possible to resolve conflicts during a pull, you may have to redo the conflict
```

resolution if your PR doesn't get merged fast enough and there are new conflicts. Do not try to
rebase the result of a `git subtree pull`, rebasing merge commits is a bad idea in general.

You always need to specify the `-P` prefix to the subtree directory and the corresponding remote
repository. If you specify the wrong directory or repository
you'll get very fun merges that try to push the wrong directory to the wrong remote repository.

> Luckily you can just abort this without any consequences and try again. It is usually fairly obvious

Reviewer 6: How do you abort?

Reviewer 5: When pulling, `git merge --abort` I'd guess? We should test it.

Pushing should always succeed, but if you try to open PR from a malformed branch, it will just look very strange and be unmergeable.

Presumably you can redo it with the right directory specified and the PR will become correct (idk if it overwrites an existing branch though, or maybe it has a `-f`/`--force` flag like `git push` does?).

that this is happening because you suddenly get thousands of commits that want to be synchronized.

#### Creating a new subtree dependency

If you want to create a new subtree dependency from an existing repository, call (from this

> repository's root directory!!)

Reviewer 6: ```suggestion
repository's root directory!)
```

Author: That was intentional :P

```
git subtree add -P src/tools/clippy https://github.com/rust-lang/rust-clippy.git master
```

This will create a new commit, which you may not rebase under any circumstances! Delete the commit
and redo the operation if you need to rebase.

Now you're done, the `src/tools/clippy` directory behaves as if clippy were part of the rustc

> monorepo, so no one but you (or others that synchronize subtrees) needs actually use `git subtree`.

Reviewer 6: ```suggestion
monorepo, so no one but you (or others that synchronize subtrees) actually need to use `git subtree`.
```

Speaking of tests, Rust has a comprehensive test suite. More information about
it can be found [here][rctd].

### External Dependencies (subtree)

As a developer to this repository, you don't have to treat the following external projects
differently from other crates that are directly in this repo:

* none so far, see https://github.com/rust-lang/rust/issues/70651 for more info

They are just regular files and directories. This is in contrast to `submodule` dependencies
(see below for those). Only tool authors will actually use any operations here.

#### Synchronizing a subtree

There are two synchronization directions: `subtree push` and `subtree pull`.

```
git subtree push -P src/tools/clippy git@github.com:your-github-name/rust-clippy sync-from-rust
```

> 

Reviewer 7: This is not working at all on my machine:

```
git subtree push -P src/tools/clippy git@github.com:matthiaskrgr/rust-clippy2  sync-from-rust

git push using:  git@github.com:matthiaskrgr/rust-clippy2 sync-from-rust
/usr/lib/git-core/git-subtree: line 757: 171266 Done                    eval "$grl"
     171267 Segmentation fault      (core dumped) | while read rev parents; do
    process_split_commit "$rev" "$parents" 0;
done
````

Author: O_o can you run a `top` in parallel and have a look at your memory consumption?

takes all the changes that
happened to the copy in this repo and creates commits on the remote repo that match the local
changes. Every local commit that touched the subtree causes a commit on the remote repo, but is
modified to move the files from the specified directory to the tool repo root.

Make sure to not pick the `master` branch on the tool repo, so you can open a normal PR to the tool

> to merge that subrepo push.

Reviewer 8: It seems to me that the important thing is to not have `git@github.com:rust-lang/rust-clippy.git` as the repository URL, i.e., you should never be pushing into branches there. But which branch you choose in your fork is largely irrelevant (you can push to master if you like).

Author: I don't have my own fork of clippy and miri, should I have?

Reviewer 3: I think it is better to keep personal feature branches out of the main repo. Not a strong opinion though.

Reviewer 5: @oli-obk Not to mention that you can only do this because of your powers on those repos, most contributors *have to* use a fork.

Reviewer: IMO for repos with many contributors the amount of feature branches in the main repo should be kept as low as possible. Clippy has even an issue, where a contributor complained about the amount of feature branches: rust-lang/rust-clippy#4745

Reviewer 7: For the record, the 
`git subtree push -P src/tools/clippy git@github.com:your-github-name/rust-clippy sync-from-rust`  command is just crashing on my machine
```
git subtree push -P src/tools/clippy git@github.com:matthiaskrgr/rust-clippy2  sync-from-rust

git push using:  git@github.com:matthiaskrgr/rust-clippy2 sync-from-rust
/usr/lib/git-core/git-subtree: line 757: 171266 Done                    eval "$grl"
     171267 Segmentation fault      (core dumped) | while read rev parents; do
    process_split_commit "$rev" "$parents" 0;
done
```

```
git subtree pull -P src/tools/clippy https://github.com/rust-lang/rust-clippy master
```

takes all changes since the last `subtree pull` from the tool repo
repo and adds these commits to the rustc repo + a merge commit that moves the tool changes into
the specified directory in the rust repository.

It is recommended that you always do a push first and get that merged to the tool master branch.
Then, when you do a pull, the merge works without conflicts.
While it's definitely possible to resolve conflicts during a pull, you may have to redo the conflict
resolution if your PR doesn't get merged fast enough and there are new conflicts. Do not try to
rebase the result of a `git subtree pull`, rebasing merge commits is a bad idea in general.

You always need to specify the `-P` prefix to the subtree directory and the corresponding remote
repository. If you specify the wrong directory or repository
you'll get very fun merges that try to push the wrong directory to the wrong remote repository.
Luckily you can just abort this without any consequences by throwing away either the pulled commits
in rustc or the pushed branch on the remote and try again. It is usually fairly obvious

> that this is happening because you suddenly get thousands of commits that want to be synchronized.

Reviewer 8: Am I correct that subtrees mean that we should no longer need to give particular priority to syncs from clippy upstream? i.e. they're "just normal PRs"? (They would essentially be no different to other PRs being made to that code).

Reviewer 6: Priority meaning `@bors p=1`? That definitely should end with the gating.

Author: jup, all the "time critical" stuff happens on the tool's repo side and not when syncing into rustc. Unless there are bug fixes, but our policy for tool bugfixes and rustc bugfixes willl not have to differ.
